README.md
# TPO project in Python
This is a pure Python project to calculate various Market profile related operations in python and visualize them interactively with the help of plotly and dash components. 
The motivation behind this project is to feed price distribution (TPOs with alphabets) and other visual cues to convolution neural networks and see if NN learns from the visualization. Therefore this project will be heavy with all the visual cues, colours and different shapes. 
While working on the project realized that it is possible to use this entire framework for regular charting as well. 

Okay, let's get started.
## Dependencies
pip install plotly (update if previously installed)

pip install dash (update if previously installed)

pip install pandas

Finally, download all the files or clone this repo as zip and extract in local folder.

## Steps for live chart

1. Open tpo_project.py from the folder in python editor

2. Run the code

3. In the console, you will see the message something like Running on http://127.0.0.1:8050/ copy the server address your console printed and paste it in your browser

4. That's it, you will get the TPO chart in browser with sample file provided in the folder.

5. About sample file: There is live.txt file in the folder. Add data manually and you should see new updated price in the chart. 

6. Marker shows the last closing price. The market color changes to green if price >= POC else it is red.

7. If the test runs successfully replace live and historical data with your desired data feed.

8. Stop the server by pressing ctrl + c in the console.

## For static chart

Simply run static_tpo.py and the chart will open in your browser. It uses a single file as a source for generating charts (unlike dash version which uses history.txt and live.txt), you can replace history.txt with your own data while making sure header names are matching. 

## Important

1. Match data format and headers exactly with given sample file.

2. It is not necessary to have the same data source for historical and live data as long as the format matches. It wouldn't matter even if symbol name is different, or there is price overlap the algorithm will drop the duplicate entries automatically.
2. Algorithm uses historical data for generating context. In tpo_project.py at the start, there is 'avglen' parameter. That number should be lower or equal to the sample size of your historical data. If you have 31 days of historical data then use avglen = 30

3. Make sure your historical file is up to date, especially entries representing day 1 and day 2. The algorithm automatically calculates session time and Initial Balance period using that info. IB is of 60 minutes. It assumes you have 1-minute data. I haven't tested on daily data. 

4. If your live feed is of 1 minute instead of a tick, you may want to increase the refresh time 'refresh_int' which is set to default 1 second.

5. Chart update will not work until live file gets at least 30 minutes worth of data. To remove the limitation reduce frequency-time 'freq' from default 30 to 5 minutes or lower. But it will affect TPO chart calculations. So if possible don't change the parameter.

## Do you have a dash compatability issue?

Run no_dash.py The flow is exactly the same as live version but without dash.

## How to read TPO chart generated by our algo and related resources

![How to read TPO chart generated by our algo-1](https://user-images.githubusercontent.com/28746824/89121455-a7b87a00-d4dc-11ea-9df1-6bc1bf897ac6.png)

![How to read charts. Hover menu -2](https://user-images.githubusercontent.com/28746824/89114961-af0d6280-d49f-11ea-8b27-baffa44c7451.png)

![How to read the charts: square boxes -3](https://user-images.githubusercontent.com/28746824/89114065-37870580-d496-11ea-99c0-a6a208441a44.png)

Highly reccomend Mind over Markets by Jim Dalton

CBOT Market Profile handbook (it is 13 mb pdf file)  https://t.co/L8DfNkLNi5?amp=1



history.txt
symbol,datetime,Open,High,Low,Close,Volume
_NF1,20200630 09:15:00,10310.0,10331.95,10310.0,10331.25,84450.0
_NF1,20200630 09:16:00,10329.15,10334.0,10314.1,10314.7,342675.0
_NF1,20200630 09:17:00,10315.0,10327.7,10313.4,10327.7,446400.0
_NF1,20200630 09:18:00,10326.0,10330.0,10320.0,10320.0,519375.0
_NF1,20200630 09:19:00,10321.0,10325.45,10316.0,10316.15,595425.0
_NF1,20200630 09:20:00,10317.0,10329.0,10315.05,10327.8,671925.0
_NF1,20200630 09:21:00,10328.2,10328.2,10311.0,10312.55,764700.0
_NF1,20200630 09:22:00,10312.5,10316.2,10308.0,10312.55,829125.0
_NF1,20200630 09:23:00,10312.45,10315.0,10305.7,10309.3,880425.0
_NF1,20200630 09:24:00,10310.0,10317.35,10306.2,10316.25,930675.0
_NF1,20200630 09:25:00,10316.2,10317.7,10312.0,10314.3,964725.0
_NF1,20200630 09:26:00,10314.35,10322.0,10314.35,10321.0,1005750.0
_NF1,20200630 09:27:00,10322.0,10322.0,10316.0,10320.0,1041375.0
_NF1,20200630 09:28:00,10320.8,10330.0,10320.8,10326.25,1125000.0
_NF1,20200630 09:29:00,10326.25,10329.5,10324.55,10328.35,1163250.0
_NF1,20200630 09:30:00,10328.35,10330.0,10324.1,10330.0,1211850.0
_NF1,20200630 09:31:00,10330.0,10339.6,10330.0,10336.1,1415100.0
_NF1,20200630 09:32:00,10335.65,10341.85,10335.65,10339.5,1482600.0
_NF1,20200630 09:33:00,10339.5,10340.65,10336.0,10337.0,1538400.0
_NF1,20200630 09:34:00,10336.85,10338.0,10333.2,10335.0,1575225.0
_NF1,20200630 09:35:00,10335.0,10335.0,10328.55,10332.55,1625625.0
_NF1,20200630 09:36:00,10330.8,10330.8,10324.1,10325.0,1658700.0
_NF1,20200630 09:37:00,10325.0,10326.95,10322.1,10324.5,1702350.0
_NF1,20200630 09:38:00,10324.1,10329.0,10322.0,10329.0,1738725.0
_NF1,20200630 09:39:00,10329.0,10333.75,10328.8,10330.0,1771500.0
_NF1,20200630 09:40:00,10329.55,10337.6,10329.55,10334.5,1800300.0
_NF1,20200630 09:41:00,10334.5,10336.0,10328.25,10329.0,1828050.0
_NF1,20200630 09:42:00,10328.05,10330.4,10326.0,10328.3,1846425.0
_NF1,20200630 09:43:00,10328.3,10330.0,10324.55,10329.0,1878825.0
_NF1,20200630 09:44:00,10329.0,10331.45,10325.3,10327.6,1927650.0
_NF1,20200630 09:45:00,10326.5,10329.0,10320.0,10322.9,1973625.0
_NF1,20200630 09:46:00,10321.55,10324.0,10313.6,10316.2,2032125.0
_NF1,20200630 09:47:00,10315.15,10323.9,10314.65,10323.9,2072775.0
_NF1,20200630 09:48:00,10322.95,10324.0,10315.95,10318.6,2111700.0
_NF1,20200630 09:49:00,10319.0,10325.05,10319.0,10325.0,2133300.0
_NF1,20200630 09:50:00,10325.75,10328.65,10322.0,10322.0,2160825.0
_NF1,20200630 09:51:00,10322.4,10322.4,10317.0,10319.05,2189700.0
_NF1,20200630 09:52:00,10318.5,10322.35,10317.15,10320.0,2205150.0
_NF1,20200630 09:53:00,10319.0,10321.95,10318.0,10321.95,2218050.0
_NF1,20200630 09:54:00,10321.8,10325.45,10318.15,10325.25,2238825.0
_NF1,20200630 09:55:00,10324.5,10327.5,10321.0,10322.7,2256900.0
_NF1,20200630 09:56:00,10323.55,10329.0,10322.0,10327.5,2278275.0
_NF1,20200630 09:57:00,10326.3,10327.6,10317.05,10317.55,2306475.0
_NF1,20200630 09:58:00,10319.55,10320.0,10310.0,10311.2,2362125.0
_NF1,20200630 09:59:00,10311.2,10314.0,10310.0,10311.95,2397375.0
_NF1,20200630 10:00:00,10312.45,10316.5,10311.3,10312.5,2431500.0
_NF1,20200630 10:01:00,10311.65,10318.4,10310.65,10316.0,2455875.0
_NF1,20200630 10:02:00,10316.0,10316.0,10311.0,10314.0,2472900.0
_NF1,20200630 10:03:00,10315.0,10317.15,10312.35,10313.05,2494575.0
_NF1,20200630 10:04:00,10314.75,10316.8,10310.0,10313.5,2530875.0
_NF1,20200630 10:05:00,10313.95,10318.9,10312.05,10316.4,2556750.0
_NF1,20200630 10:06:00,10317.4,10317.45,10312.9,10313.0,2577675.0
_NF1,20200630 10:07:00,10313.4,10314.0,10310.1,10311.35,2596950.0
_NF1,20200630 10:08:00,10311.75,10312.0,10306.65,10310.55,2658825.0
_NF1,20200630 10:09:00,10310.55,10316.8,10308.0,10315.25,2685000.0
_NF1,20200630 10:10:00,10315.2,10317.8,10314.0,10317.75,2707350.0
_NF1,20200630 10:11:00,10317.0,10324.0,10316.3,10321.5,2754900.0
_NF1,20200630 10:12:00,10321.55,10324.9,10320.0,10322.6,2775750.0
_NF1,20200630 10:13:00,10323.0,10323.9,10318.0,10319.35,2792475.0
_NF1,20200630 10:14:00,10319.0,10319.0,10314.0,10318.0,2803425.0
_NF1,20200630 10:15:00,10317.05,10324.55,10315.5,10323.25,2827950.0
_NF1,20200630 10:16:00,10324.0,10324.75,10321.0,10324.1,2848125.0
_NF1,20200630 10:17:00,10323.2,10324.95,10320.05,10323.1,2869275.0
_NF1,20200630 10:18:00,10325.0,10325.0,10320.95,10321.05,2889900.0
_NF1,20200630 10:19:00,10321.0,10322.95,10317.0,10319.0,2901300.0
_NF1,20200630 10:20:00,10319.0,10319.45,10295.0,10298.0,3176250.0
_NF1,20200630 10:21:00,10299.95,10303.0,10296.0,10299.6,3237825.0
_NF1,20200630 10:22:00,10299.6,10300.0,10293.45,10296.65,3291525.0
_NF1,20200630 10:23:00,10295.0,10299.0,10293.0,10296.0,3336300.0
_NF1,20200630 10:24:00,10298.0,10300.95,10295.0,10298.3,3372675.0
_NF1,20200630 10:25:00,10298.55,10302.45,10298.0,10300.95,3400725.0
_NF1,20200630 10:26:00,10300.05,10307.0,10300.0,10305.8,3529575.0
_NF1,20200630 10:27:00,10306.35,10307.0,10301.1,10303.0,3550650.0
_NF1,20200630 10:28:00,10303.0,10303.95,10294.05,10295.0,3587325.0
_NF1,20200630 10:29:00,10295.5,10297.0,10293.0,10297.0,3604875.0

tpo_helper.py
# -*- coding: utf-8 -*-
"""
Created on Mon Jul 27 13:26:04 2020

@author: alex1
"""
import pandas as pd
import numpy as np
import math
# import itertools


def get_ticksize(data, freq=30):
    # data = df
    numlen = int(len(data)/2)
    # sample size for calculating ticksize = 50% of most recent data
    tztail = data.tail(numlen).copy()
    tztail['tz'] = tztail.Close.rolling(freq).std()  # std. dev of 30 period rolling
    tztail = tztail.dropna()
    ticksize = np.ceil(tztail['tz'].mean()*0.25)  # 1/4 th of mean std. dev is our ticksize

    if ticksize < 0.2:
        ticksize = 0.2  # minimum ticksize limit

    return int(ticksize)


def abc(session_hr=6.5, freq=30):

    caps = [' A', ' B', ' C', ' D', ' E', ' F', ' G', ' H', ' I', ' J', ' K', ' L', ' M',
            ' N', ' O', ' P', ' Q', ' R', ' S', ' T', ' U', ' V', ' W', ' X', ' Y', ' Z']
    abc_lw = [x.lower() for x in caps]
    Aa = caps + abc_lw
    alimit = math.ceil(session_hr * (60 / freq)) + 3
    if alimit > 52:
        alphabets = Aa * int(
            (np.ceil((alimit - 52) / 52)) + 1)  # if bar frequency is less than 30 minutes then multiply list
    else:
        alphabets = Aa[0:alimit]
    bk = [28, 31, 35, 40, 33, 34, 41, 44, 35, 52, 41, 40, 46, 27, 38]
    ti = []
    for s1 in bk:
        ti.append(Aa[s1 - 1])
    tt = (''.join(ti))

    return (alphabets, tt)


def tpo(dft_rs, freq=30, ticksize=10, style='tpo', session_hr=6.5):

    if len(dft_rs) > int(60/freq):
        dft_rs = dft_rs.drop_duplicates('datetime')
        dft_rs = dft_rs.reset_index(inplace=False, drop=True)
        dft_rs['rol_mx'] = dft_rs['High'].cummax()
        dft_rs['rol_mn'] = dft_rs['Low'].cummin()
        dft_rs['ext_up'] = dft_rs['rol_mn'] > dft_rs['rol_mx'].shift(2)
        dft_rs['ext_dn'] = dft_rs['rol_mx'] < dft_rs['rol_mn'].shift(2)

        alphabets = abc(session_hr, freq)[0]
        alphabets = alphabets[0:len(dft_rs)]
        hh = dft_rs['High'].max()
        ll = dft_rs['Low'].min()
        day_range = hh - ll
        dft_rs['abc'] = alphabets
        # place represents total number of steps to take to compare the TPO count
        place = int(np.ceil((hh - ll) / ticksize))
        # kk = 0
        abl_bg = []
        tpo_countbg = []
        pricel = []
        volcountbg = []
        # datel = []
        for u in range(place):
            abl = []
            tpoc = []
            volcount = []
            p = ll + (u*ticksize)
            for lenrs in range(len(dft_rs)):
                if p >= dft_rs['Low'][lenrs] and p < dft_rs['High'][lenrs]:
                    abl.append(dft_rs['abc'][lenrs])
                    tpoc.append(1)
                    volcount.append((dft_rs['Volume'][lenrs]) / freq)
            abl_bg.append(''.join(abl))
            tpo_countbg.append(sum(tpoc))
            volcountbg.append(sum(volcount))
            pricel.append(p)

        dftpo = pd.DataFrame({'close': pricel, 'alphabets': abl_bg,
                              'tpocount': tpo_countbg, 'volsum': volcountbg})
        # drop empty rows
        dftpo['alphabets'].replace('', np.nan, inplace=True)
        dftpo = dftpo.dropna()
        dftpo = dftpo.reset_index(inplace=False, drop=True)
        dftpo = dftpo.sort_index(ascending=False)
        dftpo = dftpo.reset_index(inplace=False, drop=True)

        if style == 'tpo':
            column = 'tpocount'
        else:
            column = 'volsum'

        dfmx = dftpo[dftpo[column] == dftpo[column].max()]

        mid = ll + ((hh - ll) / 2)
        dfmax = dfmx.copy()
        dfmax['poc-mid'] = abs(dfmax['close'] - mid)
        pocidx = dfmax['poc-mid'].idxmin()
        poc = dfmax['close'][pocidx]
        poctpo = dftpo[column].max()
        tpo_updf = dftpo[dftpo['close'] > poc]
        tpo_updf = tpo_updf.sort_index(ascending=False)
        tpo_updf = tpo_updf.reset_index(inplace=False, drop=True)

        tpo_dndf = dftpo[dftpo['close'] < poc]
        tpo_dndf = tpo_dndf.reset_index(inplace=False, drop=True)

        valtpo = (dftpo[column].sum()) * 0.70

        abovepoc = tpo_updf[column].to_list()
        belowpoc = tpo_dndf[column].to_list()


        if (len(abovepoc)/2).is_integer() is False:
            abovepoc = abovepoc+[0]

        if (len(belowpoc)/2).is_integer() is False:
            belowpoc = belowpoc+[0]

        bel2 = np.array(belowpoc).reshape(-1, 2)
        bel3 = bel2.sum(axis=1)
        bel4 = list(bel3)
        abv2 = np.array(abovepoc).reshape(-1, 2)
        abv3 = abv2.sum(axis=1)
        abv4 = list(abv3)
        # cum = poctpo
        # up_i = 0
        # dn_i = 0
        df_va = pd.DataFrame({'abv': pd.Series(abv4), 'bel': pd.Series(bel4)})
        df_va = df_va.fillna(0)
        df_va['abv_idx'] = np.where(df_va.abv > df_va.bel, 1, 0)
        df_va['bel_idx'] = np.where(df_va.bel > df_va.abv, 1, 0)
        df_va['cum_tpo'] = np.where(df_va.abv > df_va.bel, df_va.abv, 0)
        df_va['cum_tpo'] = np.where(df_va.bel > df_va.abv, df_va.bel, df_va.cum_tpo)

        df_va['cum_tpo'] = np.where(df_va.abv == df_va.bel, df_va.abv+df_va.bel, df_va.cum_tpo)
        df_va['abv_idx'] = np.where(df_va.abv == df_va.bel, 1, df_va.abv_idx)
        df_va['bel_idx'] = np.where(df_va.abv == df_va.bel, 1, df_va.bel_idx)
        df_va['cum_tpo_cumsum'] = df_va.cum_tpo.cumsum()
        # haven't add poc tpo because loop cuts off way before 70% so it gives same effect
        df_va_cut = df_va[df_va.cum_tpo_cumsum + poctpo <= valtpo]
        vah_idx = (df_va_cut.abv_idx.sum())*2
        val_idx = (df_va_cut.bel_idx.sum())*2

        if vah_idx >= len(tpo_updf) and vah_idx != 0:
            vah_idx = vah_idx - 2

        if val_idx >= len(tpo_dndf) and val_idx != 0:
            val_idx = val_idx - 2

        vah = tpo_updf.close[vah_idx]
        val = tpo_dndf.close[val_idx]


        tpoval = dftpo[ticksize * 2:-(ticksize * 2)]['tpocount']  # take mid section
        exhandle_index = np.where(tpoval <= 2, tpoval.index, None)  # get index where TPOs are 2
        exhandle_index = list(filter(None, exhandle_index))
        distance = ticksize * 3  # distance b/w two ex handles / lvn
        lvn_list = []
        for ex in exhandle_index[0:-1:distance]:
            lvn_list.append(dftpo['close'][ex])

        excess_h = dftpo[0:ticksize]['tpocount'].sum() / ticksize  # take top tail
        excess_l = dftpo[-(ticksize):]['tpocount'].sum() / ticksize  # take lower tail
        excess = 0
        if excess_h == 1 and dftpo.iloc[-1]['close'] < poc:
            excess = dftpo['close'][ticksize]

        if excess_l == 1 and dftpo.iloc[-1]['close'] >= poc:
            excess = dftpo.iloc[-ticksize]['close']


        area_above_poc = dft_rs.High.max() - poc
        area_below_poc = poc - dft_rs.Low.min()
        if area_above_poc == 0:
            area_above_poc = 1
        if area_below_poc == 0:
            area_below_poc = 1
        balance = area_above_poc/area_below_poc

        if balance >= 0:
            bal_target = poc - area_above_poc
        else:
            bal_target = poc + area_below_poc

        mp = {'df': dftpo, 'vah': round(vah, 2), 'poc': round(poc, 2), 'val': round(val, 2), 'lvn': lvn_list, 'excess': round(excess, 2),
              'bal_target': round(bal_target, 2)}

    else:
        print('not enough bars for date {}'.format(dft_rs['datetime'][0]))
        mp = {}

    return mp

# !!! fetch all MP derived results here with date and do extra context analysis


def get_context(df_hi, freq=30, ticksize=5, style='tpo', session_hr=6.5):
    #    df_hi=df.copy()
    try:

        DFcontext = [group[1] for group in df_hi.groupby(df_hi.index.date)]
        dfmp_l = []
        i_poctpo_l = []
        i_tposum = []
        vah_l = []
        poc_l = []
        val_l = []
        bt_l = []
        lvn_l = []
        excess_l = []
        date_l = []
        volume_l = []
        rf_l = []
        ibv_l = []
        ibrf_l = []
        ibh_l = []
        ib_l = []
        close_l = []
        hh_l = []
        ll_l = []
        range_l = []

        for c in range(len(DFcontext)):  # c=1 for testing
            dfc1 = DFcontext[c].copy()
            dfc1.iloc[:, 2:6] = dfc1.iloc[:, 2:6].apply(pd.to_numeric)

            dfc1 = dfc1.reset_index(inplace=False, drop=True)
            mpc = tpo(dfc1, freq, ticksize, style, session_hr)
            dftmp = mpc['df']
            dfmp_l.append(dftmp)
            # for day types
            i_poctpo_l.append(dftmp['tpocount'].max())
            i_tposum.append(dftmp['tpocount'].sum())
            # !!! get value areas
            vah_l.append(mpc['vah'])
            poc_l.append(mpc['poc'])
            val_l.append(mpc['val'])

            bt_l.append(mpc['bal_target'])
            lvn_l.append(mpc['lvn'])
            excess_l.append(mpc['excess'])

            # !!! operatio of non profile stats
            date_l.append(dfc1.datetime[0])
            close_l.append(dfc1.iloc[-1]['Close'])
            ll_l.append(dfc1.High.max())
            hh_l.append(dfc1.Low.min())
            range_l.append(dfc1.High.max() - dfc1.Low.min())

            volume_l.append(dfc1.Volume.sum())
            rf_l.append(dfc1.rf.sum())
            # !!! get IB
            dfc1['cumsumvol'] = dfc1.Volume.cumsum()
            dfc1['cumsumrf'] = dfc1.rf.cumsum()
            dfc1['cumsumhigh'] = dfc1.High.cummax()
            dfc1['cumsummin'] = dfc1.Low.cummin()
            # !!! append ib values
            # 60 min = 1 hr divide by time frame to get number of bars
            ibv_l.append(dfc1.cumsumvol[int(60/freq)])
            ibrf_l.append(dfc1.cumsumrf[int(60/freq)])
            ib_l.append(dfc1.cumsummin[int(60/freq)])
            ibh_l.append(dfc1.cumsumhigh[int(60/freq)])

        # dffin = pd.concat(dfcon_l)
        max_po = max(i_poctpo_l)
        min_po = min(i_poctpo_l)

        dist_df = pd.DataFrame({'date': date_l, 'maxtpo': i_poctpo_l, 'tpocount': i_tposum, 'vahlist': vah_l,
                                'poclist': poc_l, 'vallist': val_l, 'btlist': bt_l, 'lvnlist': lvn_l, 'excesslist': excess_l,
                                'volumed': volume_l, 'rfd': rf_l, 'highd': hh_l, 'lowd': ll_l, 'ranged': range_l, 'ibh': ibh_l,
                                'ibl': ib_l, 'ibvol': ibv_l, 'ibrf': ibrf_l, 'close': close_l})

        dist_df['distr'] = dist_df.tpocount/dist_df.maxtpo
        dismean = math.floor(dist_df.distr.mean())
        dissig = math.floor(dist_df.distr.std())

        dist_df['daytype'] = np.where(np.logical_and(dist_df.distr >= dismean,
                                                     dist_df.distr < dismean + (dissig)), 'Trend Distribution Day', '')

        dist_df['daytype'] = np.where(np.logical_and(dist_df.distr < dismean,
                                                     dist_df.distr >= dismean - (dissig)), 'Normal Variation Day', dist_df['daytype'])

        dist_df['daytype'] = np.where(dist_df.distr < dismean - (dissig),
                                      'Neutral Day', dist_df['daytype'])

        dist_df['daytype'] = np.where(dist_df.distr > dismean + (dissig),
                                      'Trend Day', dist_df['daytype'])
        daytypes = dist_df['daytype'].to_list()

        # !!! get ranking based on distribution data frame aka dist_df
        ranking_df = dist_df.copy()
        ranking_df['vahtrend'] = np.where(ranking_df.vahlist >= ranking_df.vahlist.shift(), 1, -1)
        ranking_df['valtrend'] = np.where(ranking_df.vallist >= ranking_df.vallist.shift(), 1, -1)
        ranking_df['poctrend'] = np.where(ranking_df.poclist >= ranking_df.poclist.shift(), 1, -1)
        ranking_df['hhtrend'] = np.where(ranking_df.highd >= ranking_df.highd.shift(), 1, -1)
        ranking_df['lltrend'] = np.where(ranking_df.lowd >= ranking_df.lowd.shift(), 1, -1)
        ranking_df['closetrend'] = np.where(ranking_df.close >= ranking_df.close.shift(), 1, -1)
        ranking_df['cl_poc'] = np.where(ranking_df.close >= ranking_df.poclist, 1, -1)
        ranking_df['cl_vah'] = np.where(ranking_df.close >= ranking_df.vahlist, 2, 0)  # Max is 2
        ranking_df['cl_val'] = np.where(ranking_df.close <= ranking_df.vallist, -2, 0)  # Min is -2
        # !!! total 9 rankings, even though 2 of them have max score of +2 and -2 their else score set to 0 so wont exceed 100%
        ranking_df['power1'] = 100*((ranking_df.vahtrend + ranking_df.valtrend+ranking_df.poctrend+ranking_df.hhtrend +
                                     ranking_df.lltrend+ranking_df['closetrend']+ranking_df['cl_poc']+ranking_df['cl_vah']+ranking_df['cl_val'])/9)

        a, b = 70, 100
        x, y = ranking_df.power1.min(), ranking_df.power1.max()
        ranking_df['power'] = (ranking_df.power1 - x) / (y - x) * (b - a) + a

    except Exception as e:
        print(str(e))
        ranking_df = []
        dfmp_l = []

    return(dfmp_l, ranking_df)


def get_contextnow(mean_val, ranking):
    ibrankdf = ranking.copy()
    ibvol_mean = mean_val['volib_mean']
    ibrf_mean = mean_val['ibrf_mean']
    rf_mean = mean_val['rf_mean']
    vol_mean = mean_val['volume_mean']

    ibrankdf['ibmid'] = ibrankdf.ibl+((ibrankdf.ibh-ibrankdf.ibl)/2)

    ibrankdf['ib_poc'] = np.where(ibrankdf.ibmid >= ibrankdf.poclist.shift(), 1, -1)
    ibrankdf['ib_vah'] = np.where(ibrankdf.ibmid >= ibrankdf.vahlist.shift(), 2, 0)
    ibrankdf['ib_val'] = np.where(ibrankdf.ibmid <= ibrankdf.vallist.shift(), -2, 0)

    ibrankdf['ibvol_rise'] = ibrankdf.ibvol/ibvol_mean
    ibrankdf['ibvol_rise'] = ibrankdf.ibvol_rise * ibrankdf.ib_poc

    ibrankdf['power1'] = 100*((ibrankdf.ib_poc+ibrankdf.ib_vah +
                               ibrankdf.ib_val+ibrankdf.ibvol_rise)/4)

    #  normalize manually instead of sklearn minmax scaler to avoid dependency

    a, b = 50, 100
    x, y = ibrankdf.power1.min(), ibrankdf.power1.max()
    ibrankdf['power'] = (ibrankdf.power1 - x) / (y - x) * (b - a) + a

    return ibrankdf


def get_rf(df):
    df['cup'] = np.where(df['Close'] >= df['Close'].shift(), 1, -1)
    df['hup'] = np.where(df['High'] >= df['High'].shift(), 1, -1)
    df['lup'] = np.where(df['Low'] >= df['Low'].shift(), 1, -1)

    df['rf'] = df['cup'] + df['hup'] + df['lup']
    df = df.drop(['cup', 'lup', 'hup'], axis=1)
    return df


def get_mean(dfhist, avglen=30, freq=30):
    dfhist = get_rf(dfhist.copy())
    dfhistd = dfhist.resample("D").agg(
        {'symbol': 'last', 'Open': 'first', 'High': 'max', 'Low': 'min', 'Close': 'last', 'Volume': 'sum',
         'rf': 'sum', })
    dfhistd = dfhistd.dropna()
    comp_days = len(dfhistd)

    vm30 = dfhistd['Volume'].rolling(avglen).mean()
    volume_mean = vm30[len(vm30) - 1]
    rf30 = (dfhistd['rf']).rolling(avglen).mean()
    rf_mean = rf30[len(rf30) - 1]

    date2 = dfhistd.index[1].date()
    mask = dfhist.index.date < date2
    dfsession = dfhist.loc[mask]
    session_hr = math.ceil(len(dfsession)/60)
    "get IB volume mean"
    ib_start = dfhist.index.time[0]
    ib_end = dfhist.index.time[int(freq*(60/freq))]
    dfib = dfhist.between_time(ib_start, ib_end)
    # dfib = df.head(int(60/freq))
    # dfib['Volume'].plot()
    dfib = dfib.resample("D").agg(
        {'symbol': 'last', 'Open': 'first', 'High': 'max', 'Low': 'min', 'Close': 'last', 'Volume': 'sum',
         'rf': 'sum', })
    dfib = dfib.dropna()
    vib = dfib['Volume'].rolling(avglen).mean()
    volib_mean = vib[len(vib) - 1]
    ibrf30 = (dfib['rf']).rolling(avglen).mean()
    ibrf_mean = ibrf30[len(ibrf30) - 1]

    all_val = dict(volume_mean=volume_mean, rf_mean=rf_mean, volib_mean=volib_mean,
                   ibrf_mean=ibrf_mean, session_hr=session_hr)

    return all_val
